runtime/pubsub/internal/gcp/topic.go

```diff
@@ -0,0 +1,86 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/rs/zerolog"
+
+	"encore.dev/internal/ctx"
+	"encore.dev/pubsub/internal/types"
+	"encore.dev/runtime/config"
+)
+
+type topic struct {
+	client   *pubsub.Client
+	gcpTopic *pubsub.Topic
+	cfg      *config.PubsubTopic
+}
+
+func NewTopic(cfg *config.GCPProject, topicCfg *config.PubsubTopic) types.TopicImplementation {
+	// Create the topic
+	client := getClient(cfg)
+	gcpTopic := client.Topic(topicCfg.CloudName)
+
+	// Enable message ordering if we have an ordering key set
+	gcpTopic.EnableMessageOrdering = topicCfg.OrderingKey != ""
+
+	// Check we have permissions to interact with the given topic
+	// (note: the call to Topic() above only creates the object, it doesn't verify that we have permissions to interact with it)
+	_, err := gcpTopic.Config(ctx.App)
+	if err != nil {
+		panic(fmt.Sprintf("pubsub topic %s status call failed: %s", topicCfg.CloudName, err))
+	}
+
+	return &topic{client, gcpTopic, topicCfg}
+}
+
+func (t *topic) PublishMessage(ctx context.Context, attrs map[string]string, data []byte) (id string, err error) {
+	gcpMsg := &pubsub.Message{
+		Data:        data,
+		Attributes:  attrs,
+		OrderingKey: t.cfg.OrderingKey,
+	}
+
+	// Attempt to publish the message
+	return t.gcpTopic.Publish(ctx, gcpMsg).Get(ctx)
+}
+
+func (t *topic) Subscribe(logger *zerolog.Logger, _ *types.SubscriptionConfig, cfg *config.PubsubSubscription, f types.RawSubscriptionCallback) {
+	// Create the subscription object (and then check it exists on GCP's side)
+	subscription := t.client.Subscription(cfg.CloudName)
+	exists, err := subscription.Exists(ctx.App)
+	if err != nil {
+		panic(fmt.Sprintf("pubsub subscription %s for topic %s status call failed: %s", cfg.EncoreName, t.cfg.EncoreName, err))
```

Comment:
Hmm; I'm not sure I agree here.

If you're deploying a new instance, you don't want it to replace the healthy instance if some tweak or change means the new instance is unable to connect to your topics.

This panic prevents the new instance from coming up; thus keeping the old instance active. If there's a temporary GCP outage while an existing instance is running; then the loop on line 64 of this file will keep the subscription retrying until GCP is backup.

i.e. to hit this panic either GCP needs to be down _while you are deploying_ or something is wrong with your seutp.

Reply:
If GCP has a PubSub outage that causes this call to fail, this will cause any instance to crash on startup. That means autoscaling and node migrations breaks, not to mention deployments. We shouldn't bring down the whole application just because PubSub is down.

Reply:
I think this maybe is a more fundamental question about Encore deployments;

Should replica's fail to startup _if_ they are not healthy? Or should we let unhealthy replica's start assuming whatever is making them unhealthy is a transient error?

My default would be if a replica fails it's health checks, it should not be allowed to start serving traffic.

So the next question is; do we assume that being unable to subscribe to your queues is unhealthy? And if not; do we believe risking zero consumers of your queue is healthy?